internal/hardware/validator.go

```diff
@@ -63,31 +73,60 @@ func isNvme(name string) bool {
 	return strings.HasPrefix(name, "nvme")
 }
 
-func ListValidDisks(inventory *models.Inventory, minSizeRequiredInBytes int64) []*models.Disk {
-	var disks []*models.Disk
-	for _, disk := range inventory.Disks {
-		if disk.SizeBytes >= minSizeRequiredInBytes && funk.ContainsString([]string{"HDD", "SSD"}, disk.DriveType) {
-			disks = append(disks, disk)
-		}
+// DiskIsEligible checks if a disk is relevant for installation by testing
+// it against a list of predicates. Also returns all the reasons the disk
+// was found to be not eligible
+func (v *validator) DiskIsEligible(disk *models.Disk) (bool, []string) {
+	var notEligibleReasons []string
+
+	if minSizeBytes := gbToBytes(v.MinDiskSizeGb); disk.SizeBytes < minSizeBytes {
+		notEligibleReasons = append(notEligibleReasons,
+			fmt.Sprintf(
+				"Disk is too small (disk only has %s, but %s are required)",
+				humanize.Bytes(uint64(disk.SizeBytes)), humanize.Bytes(uint64(minSizeBytes))))
+	}
+
+	if allowedDriveTypes := []string{"HDD", "SSD"}; !funk.ContainsString(allowedDriveTypes, disk.DriveType) {
+		notEligibleReasons = append(notEligibleReasons,
+			fmt.Sprintf("Drive type is %s, it must be one of %s.",
+				disk.DriveType, strings.Join(allowedDriveTypes, ", ")))
 	}
 
+	eligible := len(notEligibleReasons) == 0
+
+	return eligible, notEligibleReasons
+}
+
+func (v *validator) ListValidDisks(inventory *models.Inventory) []*models.Disk {
+	eligibleDisks := funk.Filter(inventory.Disks, func(disk *models.Disk) bool {
+		// TODO: Remove this check in a future version once all inventories have been migrated
+		if !DiskEligibilityInitialized(disk) {
+			// That means that we haven't already checked the eligibility of this disk in UpdateInventory and we must check it now
+			eligible, _ := v.DiskIsEligible(disk)
+			return eligible
+		}
+
+		return disk.InstallationEligibility.Eligible
+	}).([]*models.Disk)
+
 	// Sorting list by size increase
-	sort.Slice(disks, func(i, j int) bool {
-		isNvme1 := isNvme(disks[i].Name)
-		isNvme2 := isNvme(disks[j].Name)
+	sort.Slice(eligibleDisks, func(i, j int) bool {
+		isNvme1 := isNvme(eligibleDisks[i].Name)
+		isNvme2 := isNvme(eligibleDisks[j].Name)
```

Comment:
Any idea what is going with the nvme here?
We only check the size if both disks are nvme or both arn't nvme?
@yevgeny-shnaidman ?

Reply:
we first compare types, and then we compare sizes for the disks of the same type. HDD is always before SDD ( no matter what size), SSD is always before NVME (no matter what size)

Reply:
We install on the first disk in the array, so our goal is to install on the smallest HDD??

Reply:
The idea was to install on the smallest HDD with sufficient capacity, leaving better disks for OCS.  The heuristic definitely needs adjustment.